### PR TITLE
topology: add max98390 2/4 speakers support

### PR DESCRIPTION
**Pull request by macchian**
topology: add max98390 2/4 speakers support

1. add 2ch/4ch playback configuration
2. add 98390 echo reference support
3. add bt_offload support

Signed-off-by: Mac Chiang <mac.chiang@intel.com>

### CODE REVIEW

**Review by lgirdwood — APPROVED** [review]
LGTM - I assume 1,2 & 3 have been fully validated on the relevant HW ?

File: tools/topology/CMakeLists.txt [hunk 1]
```diff
@@ -129,10 +129,12 @@ set(TPLGS
...
 	"sof-imx8mp-wm8960-mixer\;sof-imx8mp-wm8960-mixer"
 	"sof-imx8mp-wm8960-kwd\;sof-imx8mp-wm8960-kwd"
-	"sof-tgl-max98357a-rt5682\;sof-tgl-max98357a-rt5682\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DPLATFORM=tgl\;-DAMP_SSP=1"
-	"sof-tgl-max98357a-rt5682\;sof-adl-max98357a-rt5682\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DPLATFORM=adl\;-DAMP_SSP=2"
-	"sof-tgl-max98357a-rt5682\;sof-tgl-max98357a-rt5682-pdm1\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DDMIC_DAI_LINK_16k_PDM=STEREO_PDM1\;\;-DPLATFORM=tgl\;-DAMP_SSP=1"
-	"sof-tgl-max98357a-rt5682\;sof-tgl-rt1011-rt5682\;-DCODEC=RT1011\;-DFMT=s24le\;\;-DPLATFORM=tgl\;-DAMP_SSP=1"
+	"sof-tgl-max98357a-rt5682\;sof-tgl-max98357a-rt5682\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DCH=2\;-DPLATFORM=tgl\;-DAMP_SSP=1"
+	"sof-tgl-max98357a-rt5682\;sof-adl-max98357a-rt5682\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DCH=2\;-DPLATFORM=adl\;-DAMP_SSP=2"
+	"sof-tgl-max98357a-rt5682\;sof-adl-max98390-rt5682\;-DCODEC=MAX98390\;-DFMT=s32le\;-DCH=4\;-DPLATFORM=adl\;-DAMP_SSP=1"
+	"sof-tgl-max98357a-rt5682\;sof-adl-rt5682-ssp0-max98390-ssp2\;-DCODEC=MAX98390\;-DFMT=s16le\;-DCH=2\;-DPLATFORM=adl\;-DAMP_SSP=2"
```
Comment (plbossart) [inline tools/topology/CMakeLists.txt:135]: can you keep the order used previously with amp then headset codec, e.g. sof-adl-max98390-ssp2-rt5682-ssp0

File: tools/topology/CMakeLists.txt [hunk 2]
```diff
@@ -129,10 +129,12 @@ set(TPLGS
...
+	"sof-tgl-max98357a-rt5682\;sof-tgl-max98357a-rt5682-pdm1\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DCH=2\;-DDMIC_DAI_LINK_16k_PDM=STEREO_PDM1\;\;-DPLATFORM=tgl\;-DAMP_SSP=1"
+	"sof-tgl-max98357a-rt5682\;sof-tgl-rt1011-rt5682\;-DCODEC=RT1011\;-DFMT=s24le\;-DCH=2\;-DPLATFORM=tgl\;-DAMP_SSP=1"
```
Comment (plbossart) [inline tools/topology/CMakeLists.txt:137]: can you split the patches where you first modify existing topologies with new options, then add new platforms?

Comment (plbossart) [reply tools/topology/CMakeLists.txt:137]: I don't see the ADL topologies referenced in the kernel patches

sof-adl-rt5682-ssp0-max98390-ssp2.tplg

?????

**Review by plbossart — CHANGES_REQUESTED** [review]
This doesn't look very good...

File: tools/topology/CMakeLists.txt [hunk 3]
```diff
@@ -129,10 +129,12 @@ set(TPLGS
 	"sof-imx8-cs42888-mixer\;sof-imx8-cs42888-mixer"
 	"sof-imx8mp-wm8960-mixer\;sof-imx8mp-wm8960-mixer"
 	"sof-imx8mp-wm8960-kwd\;sof-imx8mp-wm8960-kwd"
-	"sof-tgl-max98357a-rt5682\;sof-tgl-max98357a-rt5682\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DPLATFORM=tgl\;-DAMP_SSP=1"
-	"sof-tgl-max98357a-rt5682\;sof-adl-max98357a-rt5682\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DPLATFORM=adl\;-DAMP_SSP=2"
-	"sof-tgl-max98357a-rt5682\;sof-tgl-max98357a-rt5682-pdm1\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DDMIC_DAI_LINK_16k_PDM=STEREO_PDM1\;\;-DPLATFORM=tgl\;-DAMP_SSP=1"
-	"sof-tgl-max98357a-rt5682\;sof-tgl-rt1011-rt5682\;-DCODEC=RT1011\;-DFMT=s24le\;\;-DPLATFORM=tgl\;-DAMP_SSP=1"
+	"sof-tgl-max98357a-rt5682\;sof-tgl-max98357a-rt5682\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DCH=2\;-DPLATFORM=tgl\;-DAMP_SSP=1"
+	"sof-tgl-max98357a-rt5682\;sof-adl-max98357a-rt5682\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DCH=2\;-DPLATFORM=adl\;-DAMP_SSP=2"
+	"sof-tgl-max98357a-rt5682\;sof-adl-max98390-rt5682\;-DCODEC=MAX98390\;-DFMT=s32le\;-DCH=4\;-DPLATFORM=adl\;-DAMP_SSP=1"
```
Comment (plbossart) [inline tools/topology/CMakeLists.txt:134]: why is the format s32le only here?

Comment (macchian) [reply tools/topology/CMakeLists.txt:134]: @plbossart , Thanks,  I should add two here
"sof-tgl-max98357a-rt5682\;sof-adl-max98390-rt5682\;-DCODEC=MAX98390\;-DFMT=s32le\;-DCH=2\;-DPLATFORM=adl\;-DAMP_SSP=1"
"sof-tgl-max98357a-rt5682\;sof-adl-max98390-rt5682-4ch\;-DCODEC=MAX98390\;-DFMT=s32le\;-DCH=4\;-DPLATFORM=adl\;-DAMP_SSP=1"

File: tools/topology/CMakeLists.txt (hunk 1 above)
Comment (macchian) [reply tools/topology/CMakeLists.txt:135]: ok.

File: tools/topology/CMakeLists.txt (hunk 2 above)
Comment (macchian) [reply tools/topology/CMakeLists.txt:137]: Do you mean split -DPLATFORM=adl to another PR?

Comment (plbossart) [reply tools/topology/CMakeLists.txt:137]: No, I meant adding the -DCH=2 macro first, then the new topology that uses DCH=4

Actually -DCH is really bad, we already use -DCHANNELS for the number of mics, so that's super confusing. Use -DNUM_AMPS instead or something more self-explanatory.

File: tools/topology/CMakeLists.txt (hunk 3 above)
Comment (macchian) [reply tools/topology/CMakeLists.txt:134]: @plbossart . As per Maxim's tdm suggestion,  it's supported 2(slot 0/1) and 4(slot 0/1/2/3) speakers.
So we verified the 2 and 4 speakers board with same topology. if no concerns, I will update new one.

if (!strcmp(codec_dai->component->name, MAX_98390_DEV0_NAME)) {
                        /* DEV0 tdm slot configuration Right */
                        snd_soc_dai_set_tdm_slot(codec_dai, 0x01, 3, 4, 32);
                }
                if (!strcmp(codec_dai->component->name, MAX_98390_DEV1_NAME)) {
                        /* DEV1 tdm slot configuration Left */
                        snd_soc_dai_set_tdm_slot(codec_dai, 0x02, 3, 4, 32);
                }

                if (!strcmp(codec_dai->component->name, MAX_98390_DEV2_NAME)) {
                        /* DEVi2 tdm slot configuration Tweeter Right */
                        snd_soc_dai_set_tdm_slot(codec_dai, 0x04, 3, 4, 32);
                }
                if (!strcmp(codec_dai->component->name, MAX_98390_DEV3_NAME)) {
                        /* DEV3 tdm slot configuration Tweeter Left */
                        snd_soc_dai_set_tdm_slot(codec_dai, 0x08, 3, 4, 32);
                }

File: tools/topology/sof-tgl-max98357a-rt5682.m4 [hunk 4]
```diff
@@ -174,6 +174,22 @@ DAI_ADD(sof/pipe-dai-playback.m4,
...
 
+ifelse(CODEC, `MAX98390', `
+# Low Latency capture pipeline 9 on PCM 6 using max 4 channels of s32le.
+# Schedule 48 frames per 1000us deadline on core 0 with priority 0
+PIPELINE_PCM_ADD(sof/pipe-passthrough-capture.m4,
+	9, 6, 4, s32le,
+	1000, 0, 0,
+	48000, 48000, 48000)
+
+# playback DAI is SSP1 using 2 periods
```
Comment (plbossart) [inline tools/topology/sof-tgl-max98357a-rt5682.m4:185]: comment says playback, code says capture?

File: tools/topology/CMakeLists.txt [hunk 5]
```diff
@@ -132,6 +132,8 @@ set(TPLGS
 	"sof-tgl-max98357a-rt5682\;sof-tgl-max98357a-rt5682\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DPLATFORM=tgl\;-DAMP_SSP=1"
 	"sof-tgl-max98357a-rt5682\;sof-adl-max98357a-rt5682\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DPLATFORM=adl\;-DAMP_SSP=2"
 	"sof-tgl-max98357a-rt5682\;sof-tgl-max98357a-rt5682-pdm1\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DDMIC_DAI_LINK_16k_PDM=STEREO_PDM1\;\;-DPLATFORM=tgl\;-DAMP_SSP=1"
+	"sof-tgl-max98357a-rt5682\;sof-adl-max98390-rt5682\;-DCODEC=MAX98390\;-DFMT=s32le\;-DPLATFORM=adl\;-DAMP_SSP=1"
+	"sof-tgl-max98357a-rt5682\;sof-adl-max98390-ssp2-rt5682-ssp0\;-DCODEC=MAX98390\;-DFMT=s32le\;-DPLATFORM=adl\;-DAMP_SSP=2"
```
Comment (plbossart) [inline tools/topology/CMakeLists.txt:136]: in your kernel comments you mentioned 3 configurations, I still see 2 here?

**Review by plbossart — CHANGES_REQUESTED** [review]

File: tools/topology/sof-tgl-max98357a-rt5682.m4 (hunk 4 above)
Comment (macchian) [reply tools/topology/sof-tgl-max98357a-rt5682.m4:185]: Thanks for pointing out. It should be capture in codes.

File: tools/topology/CMakeLists.txt (hunk 5 above)
Comment (macchian) [reply tools/topology/CMakeLists.txt:136]: By config 4 ch tdm slots that support 2 speakers(in slot0/1) and 4 speakers(slot 0/1/2/3) depends on real codec dais numbers. Or should I mentioned 2 hw boards in case of confusion?
So 
board1: SSP1 x 2 speakers (sof-adl-max98390-rt5682.tplg)
board2: SSP1 4 speakers (sof-adl-max98390-rt5682.tplg)
plus
board3: SSP2 x 2 speakers (with sof-adl-max98390-ssp2-rt5682-ssp0.tplg)

Comment (plbossart) [reply tools/topology/CMakeLists.txt:136]: Not following @macchian 

> By config 4 ch tdm slots that support 2 speakers(in slot0/1) and 4 speakers(slot 0/1/2/3) depends on real codec dais numbers. Or should I mentioned 2 hw boards in case of confusion?
> So
> board1: SSP1 x 2 speakers (sof-adl-max98390-rt5682.tplg)
> board2: SSP1 4 speakers (sof-adl-max98390-rt5682.tplg)

You have the same topology for a different configuration of the SSP1. How would you manage the clock and slot differences with the same topology file? Even if you play in stereo you need 4 slots for capture. Something's missing or I don't get your idea.

> plus
> board3: SSP2 x 2 speakers (with sof-adl-max98390-ssp2-rt5682-ssp0.tplg)

And here you mentioned SSP2 but with an -ssp0 suffix, so it's hard to follow!

Comment (macchian) [reply tools/topology/CMakeLists.txt:136]: These parts are relevant to Maxim. 
https://lore.kernel.org/all/20210405143801.29770-1-steves.lee@maximintegrated.com/

Share the 2/4 configuration table in DT from Maxim. 

Dev Name	                          Slave address	slot configuration
MX98390:02 Tweeter Right 	0x38	            maxim,vmon-slot-no = 2
		                                                            maxim,imon-slot-no = 3
MX98390:03 Tweeter Left	0x39	            maxim,vmon-slot-no = 3
		                                                            maxim,imon-slot-no = 2
MX98390:00 - Right	        0x3a	                    maxim,vmon-slot-no = 0
		                                                            maxim,imon-slot-no = 1
MX98390:01 - Left	               0x3b	                    maxim,vmon-slot-no = 1
		                                                            maxim,imon-slot-no = 0

Even if you play in stereo you need 4 slots for capture?
In stereo case, in the same tdm configuration, the echoRef recorded slot 0 and slot1 capture.

board3: SSP2 x 2 speakers (with sof-adl-max98390-ssp2-rt5682-ssp0.tplg)
I thought the -ssp0 suffix is described for rt5682? I follow the name like
"sof-tgl-max98373-rt5682\;sof-tgl-rt5682-ssp0-max98373-ssp2\;-DAMP_SSP=2"

Comment (plbossart) [reply tools/topology/CMakeLists.txt:136]: what I meant is that we use 2 slots for playback, regardless of how many amplifiers there are. But in the two amp case, we use 2 slots for capture and in the four-amp case we use 4 slots.
That means we need to use two different topologies since the SSP configuration will be different. To be clearer, ONLY the SSP configuration will be different (DAI_CONFIG)

Comment (macchian) [reply tools/topology/CMakeLists.txt:136]: @plbossart ,  sorry for confusing. yes, base on this topology here is as you thought, no matters how many amplifiers there 2 slots for playback, 2 slot for capture. 4 slots for playback , 4 slots for capture.

Comment (macchian) [reply tools/topology/CMakeLists.txt:136]: Even if you play in stereo you need 4 slots for capture.
My bad to don't answer this correctly. Yes, in stereo playback, it needs 4 slots for capture. So 

> @plbossart , sorry for confusing. yes, base on this topology here is as you thought, no matters how many amplifiers there 2 slots for playback, 2 slot for capture. 4 slots for playback , 4 slots for capture.

@plbossart my bad... please ignore my unclear comments. I answer it incorrectly. No matters how many amplifiers here 4 slots for playback and 4 slots for capture. That's the SSP1 board design.

Comment (lgirdwood) [reply tools/topology/CMakeLists.txt:136]: @macchian would be good to add these details about channel mappings due to board design in the commit message so everyone can follow.

Comment (macchian) [reply tools/topology/CMakeLists.txt:136]: Hi @plbossart , @lgirdwood , do you have further comments, good to go?

Comment (lgirdwood) [reply tools/topology/CMakeLists.txt:136]: Just a little more in the commit due to the complexity its best to state it like 
```
Board X with Codec Y uses 4 channels mapped as follows for Z (playback, capture ...)
Chan 0 = Left
Chan 1 = Right
Chan 2 = Tweeter Left
Chan 3 = Tweeter Right

And so on
```

File: tools/topology/sof-tgl-max98357a-rt5682.m4 [hunk 6]
```diff
@@ -282,6 +299,12 @@ ifelse(
 		SSP_CLOCK(fsync, 48000, codec_slave),
 		SSP_TDM(4, 25, 3, 15),
 		SSP_CONFIG_DATA(SSP, SPK_SSP_INDEX, 24)))',
+	CODEC, `MAX98390', `
+	SSP_CONFIG(DSP_B, SSP_CLOCK(mclk, 19200000, codec_mclk_in),
+		SSP_CLOCK(bclk, 6144000, codec_slave),
+		SSP_CLOCK(fsync, 48000, codec_slave),
+		SSP_TDM(4, 32, 3, 15),
```
Comment (plbossart) [inline tools/topology/sof-tgl-max98357a-rt5682.m4:306]: I am still in the dark on the configurations. On capture, the tweeter slots would be empty in the 2speaker case, so how would userspace/CRAS know which channels to use?

Do you have something like a device-specific UCM file that provides a channel map to extract the relevant channels?

You should really explain further in the commit message what the end-to-end configuration is, or assumed to be.

**Review by plbossart — CHANGES_REQUESTED** [review]
Please use the 'request review' button when you push a new version, it's not automatic and I didn't see that you pushed a new version 7 days ago.

See comment below.

File: tools/topology/sof-tgl-max98357a-rt5682.m4 (hunk 6 above)
Comment (macchian) [reply tools/topology/sof-tgl-max98357a-rt5682.m4:306]: yes, CRAS has it in UCM that provides the channel map selection.

CaptureChannelMap "0 1 -1 -1 -1 -1 -1 -1 -1 -1 -1"
or
CaptureChannelMap "0 1 2 3 -1 -1 -1 -1 -1 -1 -1"

I am trying to explain more details in commit message.  Feel free to comment if any better description.

Board A with codec max98390 uses 2 in 4 channels mapping for playback and capture.  
Chan 0 = Left
Chan 1 = Right
&
2 codecs DAI selection
'Left DAI Sel Mux' Left"
'Right DAI Sel Mux' Right"

CRAS UCM defines the CaptureChannelsMap selection.
CaptureChannelMap "0 1 -1 -1 -1 -1 -1 -1 -1 -1 -1"

Board B with codec max98390 uses 4 channels mapping for playback and capture.
Chan 0 = Left
Chan 1 = Right
Chan 2 = Tweeter Left
Chan 3 = Tweeter Right
&
4 codecs DAI selection
'Left DAI Sel Mux' Left"
'Right DAI Sel Mux' Right"
'Tweeter Left DAI Sel Mux' Left"
'Tweeter Right DAI Sel Mux' Right"

CRAS defines the CaptureChannelsMap selection.
CaptureChannelMap "0 1 2 3 -1 -1 -1 -1 -1 -1 -1"

Comment (plbossart) [reply tools/topology/sof-tgl-max98357a-rt5682.m4:306]: Suggested edits:

````
This patch adds support for three configurations:
SSP1 connects max98390 2/4 speakers
SSP2 connects max98390 2 speakers

The SSP TDM configuration uses 4 slots for playback and
4 slots for the echo reference capture - regardless of the number of speakers.

UCM files in userspace specify which channels needs to be used on the specific platform. There is no information reported by the topology/firmware related to valid channels.
````

**Review by macchian — COMMENTED** [review]
I updated the commit messages with clear description.

**Review by plbossart — APPROVED** [review]
